### ISSUE TITLE
Login with Github

### ISSUE THREAD

**Issue by roonyh** [roonyh/fossasia-gci-companion#3]
Allow users to login with Github. Should store a token for the user securely.

**Comment by roonyh** [comment]
Here are some steps

1) Understand the flow documented here in the github api docs
https://developer.github.com/v3/oauth/#non-web-application-flow

2) Show a form to get the username and password

3) Use npm package [request](https://www.npmjs.com/package/request) to make a basic authenticated request. see https://www.npmjs.com/package/request#http-authentication

4) above request should be made to https://developer.github.com/v3/oauth_authorizations/#create-a-new-authorization

5) Now we have a token and don't need to store username and password!

**Comment by roonyh** [comment]
so both @YagoGG and @abdulhannanali were interested in working on this. I think its fine both of you are working independently on this. It will be a valuable learning experience. You can share what you learn between the two of you. You will both be rewarded GCI tasks if both completed successfully. 

And you can discuss how to process further from here to store the token.

**Comment by abdulhannanali** [comment]
very cool @roonyh assign this to me please, i've claimed this task on GCI.. :+1:

**Comment by roonyh** [comment]
@abdulhannanali ok. go ahead.

**Comment by abdulhannanali** [comment]
@roonyh since electron is just essentially google chrome wrapped with node.js we can also use the web application flow which is the recommended way by github, also allows the user to revoke access at anytime. We canj just create a new browser for the user to authenticate, it will be really easy, What do you think?

**Comment by roonyh** [comment]
@abdulhannanali read through https://developer.github.com/v3/oauth/#web-application-flow and understand why it doesn't work for us. See step 2. **GitHub redirects back to your site**. A desktop application like ours does not have a **site**. So Github has no where to redirect the users. It is because of this that there is a non-web application flow. Was that clear?

**Comment by roonyh** [comment]
Users can revoke access to desktop applications too.

**Comment by abdulhannanali** [comment]
@roonyh it's entirely possible people in the PhoneGap world and the electron world do it, also after authenticating it's redirecting us to a url so we just have  to look for the redirection and get the url data which includes the code and the error after that we can request the access token from github using request, i have actually done this and I will be posting some code here in some time

**Comment by abdulhannanali** [comment]
let me just post whatever i've done till now

``` js
const qs = require('qs')
const request = require('request')
const electron = require('electron')
const BrowserWindow = electron.BrowserWindow;

  module.exports = function (options) {
  if (!options.client_id || !options.client_secret || !options.scopes) {

      throw new Error("all options not provided for oAuth Authentication")
  }


  function openWindow() {
    var authWindow = new BrowserWindow({
      width: 800,
      height: 600,
      show: false,
      'node-integration': false
    })

    var githubUrl = "https://github.com/login/oauth/authorize?"
    var authUrl = githubUrl + 'client_id=' + options.client_id + '&scope=' + options.scopes
    authWindow.loadURL(authUrl)
    authWindow.show()

    authWindow.webContents.on("will-navigate", function (event, url) {
      handleCallback(url, authWindow)
    })

    authWindow.webContents.on("did-get-redirect-request", function (event, oldURL, newURL) {
        handleCallback(newURL, authWindow)
    })

    // authWindow.on("close", function () {
    //   authWindow = null
    // }, false)

  }

  function handleCallback(url, authWindow) {
    var queryString = qs.parse(url.substring(url.indexOf("?") + 1))
    var code = queryString.code
    var error = queryString.error


    //   authWindow.destroy()
    // if (code || error) {
    // }

    if (code) {
      requestGithubToken(code)
    }

  }

  function requestGithubToken(code) {
    var tokenRequestUrl = "https://github.com/login/oauth/access_token"

    request({
      uri: tokenRequestUrl,
      formData: {
        code: code,
        client_id: options.client_id,
        client_secret: options.client_secret
      }
    }, function (error, response, body) {
      if (error || !body) {

      }
      else if (body) {
        var tokenOptions = queryString.parse(body)
        console.log(tokenOptions)
      }
    })
  }



  return {
    openWindow: openWindow
  }
}

```

**Comment by roonyh** [comment]
@abdulhannanali Ah. Right. So the users enter credentials to another electron window. Yeah this works! Thanks for bringing it up. Looking forward for a PR.

**Comment by abdulhannanali** [comment]
:+1: 

@roonyh How should i store it I am facing a problem here? I have a few options here
- chrome `localStorage` api (for this to work I will have to call the module from the browser window)
- BrowseWindow.webContents.session api of electron this will work too

I need your outpuy

**Comment by roonyh** [comment]
> chrome localStorage api (for this to work I will have to call the module from the browser window)

Which module is that? node-git? Don't worry about calling it from a browser window if thats simple.

> BrowseWindow.webContents.session api of electron this will work too

Where is it documented? Could you give me a link?

We can also use a npm module like [configstore](https://www.npmjs.com/package/configstore)

How secure is each option? Can a different application running on the machine access the token? Though I think we can store a hashed token as well.

And is the stored token loss when user quits the application?

**Comment by abdulhannanali** [comment]
@roonyh 
chrome localStorageApi is a browser api 
the token can't be accessed by any other application running on the machine, neither the client secrets, 
the electron session api is available here electron.atom.io/docs/v0.36.0/api/session/

**Comment by abdulhannanali** [comment]
@roonyh yes it's lost when we close the application, but the session between the github and the user persists and the new one is generated for the user without him entering credentials.

**Comment by roonyh** [comment]
closed by #9
